### PR TITLE
pkg/k8s: fix all structural issues with CNP validation

### LINKED ISSUE #10643: Input validation issue with CiliumNetworkPolicy
## Bug report
Bad input values in generated CiliumNetworkPolicy led to Cilium pod failures (readiness/liveness probe failures leading to Error/CrashLoop) when policy applied.

**General Information**

- Cilium version (run `cilium version`)
`# cilium version
Client: 1.6.5 88642ed70 2019-12-16T07:21:37-08:00 go version go1.12.14 linux/amd64
Daemon: 1.6.5 88642ed70 2019-12-16T07:21:37-08:00 go version go1.12.14 linux/amd64`
- Kernel version (run `uname -a`)
`# uname -r
4.19.84-coreos`
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)
`$ kubectl version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"archive", BuildDate:"2020-02-29T16:37:45Z", GoVersion:"go1.14", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.7", GitCommit:"be3d344ed06bff7a4fc60656200a93c74f31f9a4", GitTreeState:"clean", BuildDate:"2020-02-11T19:24:46Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}`
- Link to relevant artifacts (policies, deployments scripts, ...)
* Policy template: https://gist.github.com/tgelter/d123bfa6a35f524bc71bc3cc2e622984
* Generated template (which was applied, leading to the issue) https://gist.github.com/tgelter/86bb5f03adece90af9ebebf777a33eb4
- Upload a system dump (run `curl -sLO
https://github.com/cilium/cilium-sysdump/releases/latest/download/cilium-sysdump.zip &&
python cilium-sysdump.zip` and then attach the generated zip file)

**How to reproduce the issue**

1. Ensure cluster has Cilium running in a healthy state
2. Apply the bad CiliumNetworkPolicy yaml

### PR DESCRIPTION
**Pull request by aanm**
CNP schema validation was incorrectly formatted for some fields which
could cause badly formatted yaml files to be accepted by kube-apiserver
bypassing the schema validation. This would later cause Cilium to print
errors and potentially avoid it from starting, as the invalid CNPs would
prevent Cilium from fully synchronize with kube-apiserver an operation
that is essential when starting Cilium.

This commit fixes all violations presented by kubernetes for the CCNP
and CNP validation which would then deny bad CNPs and / or CCNPs from
being accepted by kube-apiserver.

Signed-off-by: André Martins <andre@cilium.io>

Fixes: #10643

```release-note
Tight CNP and CCNP schema validation for badly formatted policies (yaml or json) 
```

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

File: pkg/k8s/apis/cilium.io/v2/register.go
Comment (joestringer) [inline pkg/k8s/apis/cilium.io/v2/register.go:878]: Is there any meaningful potential impact with the case change here or are they seen as the same thing when upgrading schemas?

**Review by joestringer — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: Documentation/install/upgrade.rst
Comment (tgraf) [inline Documentation/install/upgrade.rst]: Add a sentence why this is required

File: cilium/cmd/preflight_k8s_valid_cnp.go
Comment (tgraf) [inline cilium/cmd/preflight_k8s_valid_cnp.go]: s/all/are/
... in case they are not, an error

File: cilium/cmd/preflight_k8s_valid_cnp.go
Comment (tgraf) [inline cilium/cmd/preflight_k8s_valid_cnp.go]: exit code -1 is returned.

File: cilium/cmd/preflight_k8s_valid_cnp.go
Comment (tgraf) [inline cilium/cmd/preflight_k8s_valid_cnp.go]: _K_ubernetes

**Review by tgraf — CHANGES_REQUESTED** [review]

File: cilium/cmd/preflight_k8s_valid_cnp.go
Comment (joestringer) [inline cilium/cmd/preflight_k8s_valid_cnp.go:72]: Should we have a real timeout here, and for the other contexts around this area?

**Review by joestringer — APPROVED** [review]
If a user specifies the APIserver in Helm (because of kube-proxy-free), will this tool automatically pick that up and use it?

I looked through, nothing looked overly out-of-place but I'm not the best reviewer for the k8s side.

Upgrade-wise I guess we're basically saying that the best time to validate your policy is before you move to the new version with the stricter validation?

One more question, is there an easy way for me to just run this on my cluster? Just exec into the cilium pod and run it there? (Is that advisable?) Let's say I wanted to validate my policies but without the rest of the upgrade process.

The new tool should be added to k8s code owners.

**Comment by aanm** [comment]
> If a user specifies the APIserver in Helm (because of kube-proxy-free), will this tool automatically pick that up and use it?
> 

Good point, it won't. I'll fix this in a follow up PR because this is an existing bug.

> I looked through, nothing looked overly out-of-place but I'm not the best reviewer for the k8s side.
> 
> Upgrade-wise I guess we're basically saying that the best time to validate your policy is before you move to the new version with the stricter validation?
> 

Correct.

> One more question, is there an easy way for me to just run this on my cluster? Just exec into the cilium pod and run it there? (Is that advisable?) Let's say I wanted to validate my policies but without the rest of the upgrade process.

Which policies? The ones installed in kube-apiserver will automatically be validated by kubernetes validation. Any policies the users have locally can be validated with `kubectl apply -f <file> --dry-run=server`

> 
> The new tool should be added to k8s code owners.

File: cilium/cmd/preflight_k8s_valid_cnp.go
Comment (aanm) [reply cilium/cmd/preflight_k8s_valid_cnp.go:72]: I don't think so because this is a user request and we don't know the number of CNPs installed to have a timeout here. Also, this will be triggered by the user so the user can cancel the request with CTRL+C

**Comment by aanm** [comment]
test-me-please

File: cilium/cmd/preflight_k8s_valid_cnp.go
Comment (joestringer) [reply cilium/cmd/preflight_k8s_valid_cnp.go:72]: I'm thinking from the perspective of this being run automatically as part of the preflight. What if some issue causes APIserver connectivity breakage then, will the pod just hang in not-ready state? How will users know?

**Comment by joestringer** [comment]
> > If a user specifies the APIserver in Helm (because of kube-proxy-free), will this tool automatically pick that up and use it?

> Good point, it won't. I'll fix this in a follow up PR because this is an existing bug.

Existing users could be deploying with kube-proxy free and be affected by this bug, will this tool work correctly for them during upgrade?

**Comment by aanm** [comment]
> > > If a user specifies the APIserver in Helm (because of kube-proxy-free), will this tool automatically pick that up and use it?
> 
> > Good point, it won't. I'll fix this in a follow up PR because this is an existing bug.
> 
> Existing users could be deploying with kube-proxy free and be affected by this bug, will this tool work correctly for them during upgrade?

It won't, but to fix this will require more changes that need to be done before we release the next version of 1.7 and 1.6.

**Comment by aanm** [comment]
test-ginkgo-please

**Comment by aanm** [comment]
restart-ginkgo

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
> > > > If a user specifies the APIserver in Helm (because of kube-proxy-free), will this tool automatically pick that up and use it?
> > 
> > 
> > > Good point, it won't. I'll fix this in a follow up PR because this is an existing bug.
> > 
> > 
> > Existing users could be deploying with kube-proxy free and be affected by this bug, will this tool work correctly for them during upgrade?
> 
> It won't, but to fix this will require more changes that need to be done before we release the next version of 1.7 and 1.6. Which I will take care of it before we do the release.

I've managed to add this suggestion without changing any code. Please check the documentation changes.

**Review by joestringer — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Review by aanm — COMMENTED** [review]
There was a typo in the docs:
```patch
diff --git b/Documentation/install/upgrade.rst a/Documentation/install/upgrade.rst
index fb35dabae..2d4d02bac 100644
--- b/Documentation/install/upgrade.rst
+++ a/Documentation/install/upgrade.rst
@@ -66,7 +66,7 @@ file.
         --set config.enabled=false \\
         --set operator.enabled=false \\
         --set global.k8sServiceHost=API_SERVER_IP \\
-        --set global.k8sServicePort=API_SERVER_PORT
+        --set global.k8sServicePort=API_SERVER_PORT \\
         > cilium-preflight.yaml
       kubectl create cilium-preflight.yaml
```

I've fix it. Will merge the PR once the GH check for docs have passed. All other tests have passed before this typo was fixed.

**Pull request merged**
